### ISSUE TITLE
[BLUEMARLIN-29] : For DIN-Lookalike model, training is very slow. #52

### ISSUE THREAD

**Issue by srinaginikonduru** [Futurewei-io/blue-marlin#261]
[Bimlesh759-AI](https://github.com/Bimlesh759-AI)

Training scenario:

   Following datasets details include users with minimum one click count with step = 10.
   test_dataset_count = 110755727,
   train_dataset_count = 517801469,
   user_count = 94315979,
   item_count = 19
   
   EPOCH = 250
   train_batch_size = 20480
       test_batch_size = 2048
   
   Current model takes around 12 hours to train 1 epoch if we use all datasets. If we use around 50% datasets by randomly selecting
   then also model takes around 7-8 hours to train for 1 epoch.
   
   By this analogy, If we want to train the model for complete 250 epochs on full datasets, then it will take around 125 days.
   
   Currently we are using Tensorflow 1.15, Two GPU are there in training but only one GPU is used.

Target about model.

   1. It is required that model should not take more than 24 hours to train.
   2. Model should be able to use all the available GPU.
   3. Is it possible to further reduce the datasets with regard to size without losing insights.
   4. Is it possible to get DIN-Lookalike model and trainer code in Tensorflow 2.0 version.

**Comment by srinaginikonduru** [comment]
[jimmylao](https://github.com/jimmylao)



@Bimlesh759-AI
More data does not always generate better model. High quality data usually generates better model.
Most of the time, real world data is noisy and there exists a lot of redundancy. You may need to have more insight of the data.

1.Break-down information regarding your training/testing data? i.e.
-the number of positive and negative training/testing samples of all 19 items
-the number of positive and negative training/testing samples of each of the 19 items
you may want to balance the number of training/testing samples of these 19 items.
2.Since your dataset is huge, it's quite essential to do parallel training. I think you may try to use TF 2 for multiple GPU training already, how's your progress on this topic? There's another option: TF 1 + horovod, to my understanding, you tried this approach too, can you sync up what is the current status of parallel training with multiple GPUs or machines?

**Comment by srinaginikonduru** [comment]
[Bimlesh759-AI](https://github.com/Bimlesh759-AI)


Details about training datasets:
Positive labels - 217949105
Negative labels- 233029477

total number of rows in datasets- 450978582
total number of rows in datasets with unique aid_index (or aid)- 91152601

Below table shows no. of rows (count) with particular keyword id.
+-------+---------+
|keyword| count|
+-------+---------+
| 16| 13850541|
| 17| 6065903|
| 7| 10124720|
| 14| 14468395|
| 12|102332499|
| 8| 14433430|
| 10| 22725438|
| 11| 7802683|
| 2| 3771215|
| 3| 9103754|
| 5| 6555940|
| 15| 18514390|
| 9| 8764869|
| 0| 12021109|
| 1|127750503|
| 13| 2956558|
| 6| 52615187|
| 4| 17121448|
+-------+---------+

Details about test datasets:
Positive labels - 18131220
Negative labels- 48307204

total number of rows in datasets-66438424
total number of rows in datasets with unique aid_index (or aid)- 30294628

Below table shows no. of rows (count) with particular keyword id.
+-------+--------+
|keyword| count|
+-------+--------+
| 16| 2127218|
| 17| 1646675|
| 7| 2001651|
| 14| 2309676|
| 12|11349194|
| 8| 2383793|
| 10| 3709643|
| 11| 1878372|
| 2| 949824|
| 3| 1919426|
| 5| 1417523|
| 15| 2794550|
| 9| 2006817|
| 0| 2135362|
| 1|16328372|
| 13| 950384|
| 6| 7033431|
| 4| 3496513|
+-------+--------+
1.
Above is the details regarding the training and test datasets. Please suggest if it is possible to reduce it without compromising on model accuracy in future.

2.Currently we have not tried TF2.0 for multiple gpu but implementation for multiple gpu is underway using TF1 + horovod

**Comment by srinaginikonduru** [comment]
<html>
<body>
<p dir="auto" style="box-sizing: border-box; margin-top: 0px !important; margin-bottom: 16px; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a class="user-mention" data-hovercard-type="user" data-hovercard-url="/users/Bimlesh759-AI/hovercard" data-octo-click="hovercard-link-click" data-octo-dimensions="link_type:self" href="https://github.com/Bimlesh759-AI" style="box-sizing: border-box; background-color: transparent; color: var(--color-fg-default); text-decoration: none; font-weight: 600; white-space: nowrap;">@Bimlesh759-AI</a><span> </span>Thanks for sharing.<br style="box-sizing: border-box;">As shown in the table below, the ratios between positive and negative samples are 1.07 in training data and 2.66 in testing data. These are balanced (good) ratios.</p>

  | positive # | negative # | pos vs neg ratio | %
-- | -- | -- | -- | --
train | 217,949,105 | 233,029,477 | 1.069192172 | 87.16%
test | 18,131,220 | 48,307,204 | 2.66431073 | 12.84%

<p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 16px; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">One approach you may want to try is: keep all testing samples and randomly drop 50% of training samples, then train a model to see if there's significant drop in AUC for testing samples. (Since the ratio of testing samples is 13%-, it has much less contribution to training time).</p><p dir="auto" style="box-sizing: border-box; margin-top: 0px; margin-bottom: 0px !important; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Based on your keyword break-down sample information, it looks good too. The following are curves of accumulated percentage of training and testing samples with all keywords. It shows that keyword samples agrees well in training and testing.<br style="box-sizing: border-box;"></p>
</body>
</html>

<img width="545" alt="sample" src="https://user-images.githubusercontent.com/101370972/157877130-4b6f0230-ef3a-4ffd-a789-895e191ffa7c.PNG">

**Comment by srinaginikonduru** [comment]
<html>
<body>
<p dir="auto" style="box-sizing: border-box; margin-top: 0px !important; margin-bottom: 16px; color: rgb(36, 41, 47); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Some quick investigation shows significant speed-up with multiple GPU using TF2, however, as shown in the curve below, the speedup is not linear to the number of GPUs used, there is a balance between GPU # vs speed-up efficiency. The efficiency drops quickly after using 4 GPUs.</p>

GPU # | batch size | time (seconds) | Speed-up
-- | -- | -- | --
1 | 4,096 | 572 | 1
2 | 8,192 | 300 | 1.90667
3 | 12,288 | 224 | 2.55357
4 | 16,384 | 195 | 2.93333
5 | 20,480 | 176 | 3.25
6 | 24,576 | 168 | 3.40476
7 | 28,672 | 164 | 3.4878
8 | 32,768 | 162 | 3.53086


</body>
</html>
<img width="451" alt="speed-up" src="https://user-images.githubusercontent.com/101370972/157877513-396bc0c3-615a-4284-a0ac-84cc41756b7d.PNG">

**Comment by srinaginikonduru** [comment]
[Bimlesh759-AI](https://github.com/Bimlesh759-AI)

Update on Horovod Implementation
Below is the few lines of log. Currently 2 GPUs are used, using horovod implementation.
Around 8-9 hours will be taken in one complete epoch for half of total training datasets.

2022-03-07T14:37:31.14694774Z Current epoch is :0 out of total epoch size : 250
2022-03-07T14:37:31.151764703Z Current epoch is :0 out of total epoch size : 250
2022-03-07T14:37:34.149059123Z 2022-03-07 22:37:34.147682: Successfully opened dynamic library libcublas.so.10.0
2022-03-07T14:37:34.195508001Z 2022-03-07 22:37:34.194611: Successfully opened dynamic library libcublas.so.10.0
2022-03-07T14:49:41.36024786Z Current batch is :500 out of total batch size : 25284
2022-03-07T14:49:41.360453056Z Current GPU is 0
2022-03-07T14:49:53.371842581Z Current batch is :500 out of total batch size : 25284
2022-03-07T14:49:53.371892941Z Current GPU is 1
2022-03-07T15:00:15.9504008Z Current batch is :1000 out of total batch size : 25284
2022-03-07T15:00:15.950470233Z Current GPU is 1
2022-03-07T15:00:16.496543414Z Current batch is :1000 out of total batch size : 25284
2022-03-07T15:00:16.496598614Z Current GPU is 0
2022-03-07T15:12:15.966611817Z Current batch is :1500 out of total batch size : 25284
2022-03-07T15:12:15.966671727Z Current GPU is 0
2022-03-07T15:13:11.603205931Z Current batch is :1500 out of total batch size : 25284
2022-03-07T15:13:11.603255662Z Current GPU is 1
2022-03-07T15:26:53.688137055Z Current batch is :2000 out of total batch size : 25284
2022-03-07T15:26:53.688267245Z Current GPU is 0

**Comment by srinaginikonduru** [comment]
[jimmylao](https://github.com/jimmylao)


@Bimlesh759-AI Thanks for the update. Do you have information regarding

How long does one epoch take using single GPU with all training samples?
How much speed-up of 2 GPUs with Horovod vs 1 GPU without Horovod? Since one epoch will take very long time, you may have an estimation, say: the time used by 1 GPU and 2 GPUs for the first 1000 batches.
